### PR TITLE
systran support pre-request

### PR DESCRIPTION
**Pull request by fondbcn**
## Checklist

issue:#10799

- [ ] Lint and unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added documentation to describe my feature.
- [ ] I have squashed my commits into logic units.
- [x] I have described the changes in the commit messages.

## Other information

This PR is not to approve directly since I have an issue with systran api, which gives 403 error (IP_ADDRESS_RESTRICTION)

### CODE REVIEW

**Comment by nijel** [comment]
Can you please add tests and documentation for this as well_

**Comment by fondbcn** [comment]
> Can you please add tests and documentation for this as well?

it is not tested, bcz the "https://api-translate.systran.net/" gives 403 ip_restriction and the tutorial shows ip gate instead of this api link ! "https://www.youtube.com/watch?v=QILC9eOCjDw&t=569s"
Can you please test the api link and give me some orientations ?

**Comment by nijel** [comment]
For testing we mock the API see for example:

https://github.com/WeblateOrg/weblate/blob/31358338c3613433794d19308c86269514232eda/weblate/machinery/tests.py#L1381-L1414

But you should also test against the real service, you can get a free API key after registering with them.

**Comment by fondbcn** [comment]
tests.py :
`class SystranTranslationTest(BaseMachineTranslationTest):
    MACHINE_CLS = SystranTranslation
    EXPECTED_LEN = 1
    CONFIGURATION = {
        "key": "key",
    }

    def mock_empty(self):
        raise SkipTest("Not tested")

    def mock_error(self):
        raise SkipTest("Not tested")

    def mock_response(self):
        responses.add(
            responses.GET,
            "https://api-translate.systran.net/translation/apiVersion",
            json={"version":"2.11.0"},
        )
        
        responses.add(
            responses.GET,
            "https://api-translate.systran.net/translation/supportedLanguages",
            json={
                "languagePairs":[
                    {"source":"ar","target":"cs","profiles":[
                        {"id":"cde18bf3-a0ff-42ec-92bd-147e695f358e","private":False,"selectors":{
                            "domain":"Generic","owner":"Systran","size":"L","tech":{
                                "name":"Docker-OpenNMT-ctranslate","type":"NMT"
                            },
                        },}
                    ]},
                ],
            }
        )
        
        responses.add(
            responses.GET,
            "https://api-translate.systran.net/translation/text/translate",
            json={"outputs":[{"output":"Hallo"}]},
        )`
Error :
`self.assertGreater(unit.machinery["quality"][0], -1)
KeyError: 'quality'`
Fail :
`self.assertTrue(machine_translation.is_supported(self.ENGLISH, self.SUPPORTED))
AssertionError: False is not true`

Is that from migration ? if yes, I don't know how to deal with "missing version to migrate from!"

**Comment by nijel** [comment]
>  self.assertTrue(machine_translation.is_supported(self.ENGLISH, self.SUPPORTED)) AssertionError: False is not true

You only mock `ar`-`cs` as a supported language pair, while this tests whether translation from English to given language is supported.

>  self.assertGreater(unit.machinery["quality"][0], -1) KeyError: 'quality'

This is probably the outcome of that as well - it produces no translations here.

**Comment by fondbcn** [comment]
> You only mock `ar`-`cs` as a supported language pair, while this tests whether translation from English to given language is supported.


The same failure and error for "en".
Is migration for "weblate.machinery.models" needed ? related to that ? how to do it ("missed version to migrate from") ?
```
SYSTRAN_LANGUAGE_JSON={
    "languagePairs":[
        {
            "source":"en","target":"cs","profiles":[
                {
                    "id":"32e871bd-c82f-4b36-a59f-7cfd109a606e","private":False,"selectors":{
                        "domain":"Generic","owner":"Systran","size":"L","tech":{
                            "name":"OpenNMT-ctranslate","type":"NMT"
                        }
                    }
                },
                {
                    "id":"4c02852a-54f4-4c26-81c1-271c71fea810","private":False,"selectors":{
                        "domain":"Cybersecurity","owner":"Systran","size":"L","tech":{
                            "name":"OpenNMT-ctranslate","type":"NMT"
                        }
                    }
                },
            ],
        },
    ],
}
class SystranTranslationTest(BaseMachineTranslationTest):
    MACHINE_CLS = SystranTranslation
    EXPECTED_LEN = 1
    CONFIGURATION = {
        "key": "key",
    }

    def mock_empty(self):
        raise SkipTest("Not tested")

    def mock_error(self):
        raise SkipTest("Not tested")

    def mock_response(self):
        responses.add(
            responses.GET,
            "https://api-translate.systran.net/translation/apiVersion",
            json={"version":"2.11.0"},
        )
        
        responses.add(
            responses.GET,
            "https://api-translate.systran.net/translation/supportedLanguages",
            json=SYSTRAN_LANGUAGE_JSON
        )
        
        responses.add(
            responses.GET,
            "https://api-translate.systran.net/translation/text/translate",
            json={"outputs":[{"output":"ahoj"}]},
        )
```
is this "key": "key" correct ?

**Comment by fondbcn** [comment]
> Can you please include these in the pull request so that I can comment on it and see the actual test results?

Ran 7 tests in 0.051s
FAILED (failures=1, errors=1, skipped=2)

File: weblate/settings_test.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
 ADMINS = (("Weblate test", "noreply@weblate.org"),)
 
 # The secret key is needed for tests
-SECRET_KEY = "secret key used for tests only"  # noqa: S105
+SECRET_KEY = "4bc968d5-ee92-42de-8eff-cfa315606f34"  # noqa: S105
```
Comment (nijel) [inline weblate/settings_test.py:49]: This doesn't look like an intentional change.

File: weblate/machinery/tests.py [hunk 2]
```diff
@@ -904,6 +934,39 @@ def test_bug(self):
...
+        )
+
+        responses.add(
+            responses.GET,
+            "https://api-translate.systran.net/translation/supportedLanguages",
+            json=SYSTRAN_LANGUAGE_JSON,
+        )
+
+        responses.add(
+            responses.GET,
```
Comment (nijel) [inline weblate/machinery/tests.py:964]: POST is used here (see https://github.com/WeblateOrg/weblate/pull/10955/files#diff-21a569b0906c395b0a03c3c2d6499eac7101930e86d70ef9a5df9937a66e6899R43):

```suggestion
            responses.POST,
```

File: weblate/machinery/systran.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    def download_languages(self):
+        """Download list of supported languages from a service."""
+        response = self.request(
+            "get",
+            "https://api-translate.systran.net/translation/supportedLanguages",
+            params={"key": self.settings["key"]},
+        )
+        payload = response.json()
+        self.check_failure(payload)
+        return [pair["target"] for pair in payload["languagePairs"]]
```
Comment (nijel) [inline weblate/machinery/systran.py:30]: You only extract `target` from the supported pairs. That's why it doesn't recognize pair as supported in the test. You should fetch full pairs and adjust `is_supported` method to work on it. You can basically copy an implementation from Apertium:

https://github.com/WeblateOrg/weblate/blob/18d07a1fb548cb41a30ee0c9d02e8bfcc874d74a/weblate/machinery/apertium.py#L91-L101

File: weblate/machinery/systran.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        source,
+        language,
+        text: str,
+        unit,
+        user,
+        threshold: int = 75,
+    ):
+        """Download list of possible translations from a service."""
+        response = self.request(
+            "get",
```
Comment (nijel) [inline weblate/machinery/systran.py:47]: The [API documentation](https://docs.systran.net/translateAPI/translation/#tag-Translation) says this should be POST.

```suggestion
            "post",
```

Needs a test mock change already suggested before.

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Pull request merged**